src/gitlab-api/index.ts

```diff
@@ -0,0 +1,153 @@
+import { XMLParser } from 'fast-xml-parser'
+import {
+    WaylandElementType,
+    WaylandProtocol as WaylandProtocolModel,
+} from '../model/wayland'
+import { transformXMLElement } from '../lib/xml-protocol-transformers'
+
+export interface GitLabFile {
+    name: string
+    webPath: string
+    path: string
+    protocol: WaylandProtocolModel
+}
+
+export interface GitLabMergeRequest {
+    iid: number
+    title: string
+    diffStats: { path: string }[]
+    diffHeadSha: string
+}
+
+export interface GitLabPaginationInfo {
+    endCursor: string
+    hasNextPage: boolean
+}
+
+export interface GetMergeRequestsResponse {
+    pageInfo: GitLabPaginationInfo
+    mrs: GitLabMergeRequest[]
+}
+
+async function callGraphQl(query: string): Promise<any> {
+    const body = JSON.stringify({ query })
+    return await fetch('https://gitlab.freedesktop.org/api/graphql', {
+        method: 'post',
+        body,
+        headers: new Headers({
+            'Content-Type': 'application/json',
+            'Content-Length': body.length.toString(),
+        }),
+    })
+}
+
+function xml_filter({ path }: { path: string }) {
+    return path.endsWith('.xml')
+}
+
+export async function getMergeRequestFiles(
+    mr: GitLabMergeRequest
+): Promise<GitLabFile[]> {
+    const paths = mr.diffStats.map(({ path }) => path)
+    const response = await callGraphQl(`{
+      project(fullPath: "wayland/wayland-protocols") {
+        repository {
+          blobs(ref: "${mr.diffHeadSha}", paths: ${JSON.stringify(paths)}) {
+            nodes {
+              name
+              webPath
+              path
+              rawBlob
+            }
+          }
+        }
+      }
+    }`)
+
+    const nodes: {
+        name: string
+        webPath: string
+        path: string
+        rawBlob: string
+    }[] = (await response.json()).data.project.repository.blobs.nodes
+
+    return nodes.reduce<GitLabFile[]>((out, node) => {
+        const parser = new XMLParser({
+            ignoreAttributes: false,
+            attributeNamePrefix: '',
```

Comment:
This is the same code from `convert-wayland-xml-to-json.ts` that's needed for lists to display correctly. I'm not sure if there's a MR that uses lists to test it.

Actually it's also missing in `regenerate-protocols-data.ts` (I plan to rework both scripts to share most code after the cross-protocol link changes).

```suggestion
            attributeNamePrefix: '',
            tagValueProcessor: (
                tagname,
                tagValue,
                jPath,
                hasAttributes,
                isLeadNode
            ) => {
                if (!isLeadNode || typeof tagValue !== 'string') return null
                return tagValue
                    .split('\n')
                    .map((line) => line.trim())
                    .join('\n')
            },
```

Reply:
This code already exists on main in GitLab.ts, I just moved it around, so if there is something wrong with it let's just fix it on main and I will rebase accordingly. I don't even understand the suggested snipit so I would rather not do this in this MR.

Reply:
Sounds good, I'll make a separate PR after this is merged I think I'll be less hassle than rebasing.

What is does is remove extra indentation around lines in descriptions (and technically summaries but they are one line and values are already trimmed) so you don't get `\n\t` instead of `\n` between lines. Normally this isn't important because html mostly ignores whitespace, but for parsing lists looks for `\n*`.